compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/inlineclasses/InlineClassManager.kt

```diff
@@ -0,0 +1,145 @@
+                !it.hasInlineClassParameters || it.isSyntheticInlineClassMember || it.origin == IrDeclarationOrigin.LOCAL_FUNCTION_FOR_LAMBDA -> null
+                it.hasMethodReplacement -> createMethodReplacement(it)
+                it.hasStaticReplacement -> createStaticReplacement(it)
+                else -> null
+            }
+        }
+
+    /**
+     * Get the box function for an inline class. Concretely, this is a synthetic
+     * static function named "box-impl" which takes an unboxed value and returns
+     * a boxed value.
+     */
+    val getBoxFunction: (IrClass) -> IrSimpleFunction =
+        storageManager.createMemoizedFunction { irClass ->
+            require(irClass.isInline)
+            buildFun {
+                name = Name.identifier("box-impl")
+                origin = JvmLoweredDeclarationOrigin.SYNTHETIC_INLINE_CLASS_MEMBER
+                returnType = irClass.defaultType
+            }.apply {
+                parent = irClass
+                copyTypeParametersFrom(irClass)
+                addValueParameter {
+                    name = Name.identifier("v")
+                    type = InlineClassAbi.getUnderlyingType(irClass)
+                }
+            }
+        }
+
+    /**
+     * Get the unbox function for an inline class. Concretely, this is a synthetic
+     * member function named "unbox-impl" which returns an unboxed result.
+     */
+    val getUnboxFunction: (IrClass) -> IrSimpleFunction =
+        storageManager.createMemoizedFunction { irClass ->
+            require(irClass.isInline)
+            buildFun {
+                name = Name.identifier("unbox-impl")
+                origin = JvmLoweredDeclarationOrigin.SYNTHETIC_INLINE_CLASS_MEMBER
+                returnType = InlineClassAbi.getUnderlyingType(irClass)
+            }.apply {
+                parent = irClass
+                createDispatchReceiverParameter()
+            }
+        }
+
+    private fun createMethodReplacement(function: IrFunction): IrReplacementFunction? {
+        require(function.dispatchReceiverParameter != null && function is IrSimpleFunction)
+        val overrides = function.overriddenSymbols.mapNotNull {
+            getReplacementFunction(it.owner)?.function?.symbol as? IrSimpleFunctionSymbol
+        }
+        if (function.origin == IrDeclarationOrigin.FAKE_OVERRIDE && overrides.isEmpty())
+            return null
+
+        val parameterMap = mutableMapOf<IrValueParameterSymbol, IrValueParameter>()
+        val replacement = buildReplacement(function) {
+            annotations += function.annotations
+            metadata = function.metadata
+            overriddenSymbols.addAll(overrides)
+
+            for ((index, parameter) in function.explicitParameters.withIndex()) {
+                val name = if (parameter == function.extensionReceiverParameter) Name.identifier("\$receiver") else parameter.name
+                val newParameter = parameter.copyTo(this, index = index - 1, name = name)
```

Comment:
Why `index - 1`?

Reply:
This method is only called on functions with non-zero dispatch receiver. Thus the iterator for `function.explicitParameters.withIndex()` starts with the dispatch receiver at index `0`, while it should get index `-1`. At the same time we are turning the extension receiver into a normal value parameter, which should be at index `0` instead of `-1`. This is why we can't just use `parameter.index` and instead have to use `index - 1`.